sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcSerializer.scala

```diff
@@ -150,69 +156,110 @@ class OrcSerializer(dataSchema: StructType) {
+//      if (reuseObj) {
+//        val result = new OrcList[WritableComparable[_]](orcSchema)
+//        (getter, ordinal) => {
+//          result.clear()
+//          baseArrayConverter.apply(result, getter, ordinal)
+//          result
+//        }
+//      } else {
+        (getter, ordinal) => {
+          val result = new OrcList[WritableComparable[_]](orcSchema)
+          baseArrayConverter.apply(result, getter, ordinal)
+          result
+        }
+//      }
+
+    case MapType(keyType, valueType, _) =>
+      val orcSchema = schemaConverter(dataType)
       // Need to put all converted values to a list, can't reuse object.
-      val keyConverter = newConverter(keyType, reuseObj = false)
-      val valueConverter = newConverter(valueType, reuseObj = false)
-      val map = getter.getMap(ordinal)
-      val keyArray = map.keyArray()
-      val valueArray = map.valueArray()
-      var i = 0
-      while (i < map.numElements()) {
-        val key = keyConverter(keyArray, i)
-        if (valueArray.isNullAt(i)) {
-          result.put(key, null)
-        } else {
-          result.put(key, valueConverter(valueArray, i))
+      val keyConverter = newConverter(keyType, false)
+      val valueConverter = newConverter(valueType, false)
+      val baseMapConverter:
+        (OrcMap[WritableComparable[_], WritableComparable[_]], SpecializedGetters, Int) => Unit = {
+        (result, getter, ordinal) => {
+          val map = getter.getMap(ordinal)
+          val keyArray = map.keyArray()
+          val valueArray = map.valueArray()
+          var i = 0
+          while (i < map.numElements()) {
+            val key = keyConverter(keyArray, i)
+            if (valueArray.isNullAt(i)) {
+              result.put(key, null)
+            } else {
+              result.put(key, valueConverter(valueArray, i))
+            }
+            i += 1
+          }
         }
-        i += 1
       }
-      result
 
-    case udt: UserDefinedType[_] => newConverter(udt.sqlType)
+//      if (reuseObj) {
+//        val result = new OrcMap[WritableComparable[_], WritableComparable[_]](orcSchema)
+//        (getter, ordinal) => {
+//          result.clear()
+//          baseMapConverter.apply(result, getter, ordinal)
+//          result
+//        }
+//      } else {
+        (getter, ordinal) => {
+          val result = new OrcMap[WritableComparable[_], WritableComparable[_]](orcSchema)
+          baseMapConverter.apply(result, getter, ordinal)
+          result
+        }
+//      }
+
+    case udt: UserDefinedType[_] => newConverter(udt.sqlType, reuseObj)
 
     case _ =>
       throw new UnsupportedOperationException(s"$dataType is not supported yet.")
   }
-
-  /**
-   * Return a Orc value object for the given Spark schema.
-   */
-  private def createOrcValue(dataType: DataType) = {
-    OrcStruct.createValue(TypeDescription.fromString(OrcFileFormat.getQuotedSchemaString(dataType)))
-  }
```

Comment:
From the profiler, this was a prominent hotspot. There are two problems here.
1) Converts dataType => String => orc type description for each data point. Instead can be done only once at the schema level with the new method schemaConverter
2) OrcStruct.createValue (https://github.com/apache/orc/blob/master/java/mapreduce/src/java/org/apache/orc/mapred/OrcStruct.java) for a struct type would recursively go down the fields and create relevant objects until it hits either a map, array, union, or a primitive type. We weren't actually using these child objects here, so now with the change it will only create the one OrcStruct object that is needed.